### ISSUE TITLE
i18n-header is parsed

### ISSUE THREAD

**Issue by cekageka** [untra/polyglot#20]
Hi,
am I doing something wrong or is it a bug: My i18n-headers look 'double-parsed', see yourself: 

```
<meta http-equiv="Content-Language" content="de">
<link rel="alternate" i18n="en" href="/de/imprint/" />
<link rel="alternate" i18n="de" href="/de/de/imprint/" />
```

Thanks,
Christian

**Comment by untra** [comment]
That does look double parsed. Can I see your jekyll code? hreflang alternate code's should be using the external url of your website, and should look more like this:

```html
<meta http-equiv="Content-Language" content="de">
<link rel="alternate" i18n="en" href="https://mywebsite.com/imprint/" />
<link rel="alternate" i18n="de" href="https://mywebsite.com/de/imprint/" />
```

**Comment by untra** [comment]
Also, what version of polyglot are you using?

**Comment by cekageka** [comment]
You're right, I removed the #url for testing purposes. Beside that it's a 1:1 copy of the current GH master branch.

**Comment by cekageka** [comment]
I have re-checke, it's 100% the current code, it's just url is empty in the dev-version I am currently working on.

**Comment by untra** [comment]
can you show me the code? Otherwise I really can't help you, haha :smile:

**Comment by cekageka** [comment]
Sure, but like I said, it's yours ;-)

```
module Jekyll
  module Polyglot
    module Liquid
      class I18nHeadersTag < ::Liquid::Tag
        def initialize(tag_name, text, tokens)
          super
          @url = text
          @url.strip!
          @url.chomp! '/'
        end

        def render(context)
          site = context.registers[:site]
          permalink = context.registers[:page]['permalink']
          i18n = "<meta http-equiv=\"Content-Language\" content=\"#{site.active_lang}\">\n"
          i18n += "  <link rel=\"alternate\" i18n=\"#{site.default_lang}\" href=\"#{@url}#{permalink}\" />\n"
          site.languages.each do |lang|
            next if lang == site.default_lang
            i18n += "  <link rel=\"alternate\" i18n=\"#{lang}\" href=\"#{@url}/#{lang}#{permalink}\" /> "
          end
          i18n
        end
      end
    end
  end
end

Liquid::Template.register_tag('I18n_Headers', Jekyll::Polyglot::Liquid::I18nHeadersTag)
```

What I assume is, that the link is changed to i.e. '/de/page' and your main code is later changing the /page to /de/page because of the language link processing. The result is /de/de/page...

**Comment by untra** [comment]
I'm sorry, let me be clearer: I want to see your jekyll repo, and in particular how you are using the `{%I18n_headers %}` tag. Is that cool? or is it a private repo?

**Comment by cekageka** [comment]
I see, but I am afraid I can't do that. It's in a private repro in GitLab. Well, if you're sure this works well I may have to build a short Github repro in the next days. My current repro is still quite fresh, I don't see much specials wich could be resonsible for the problem.

But if it may help, here's the header:

```
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  {% if jekyll.environment != "production" %}<meta name="robots" content="noindex, nofollow">{% endif %}
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta http-equiv="content-type" content="text/html; charset=utf-8">
  {% I18n_Headers %}
  <link rel="icon" type="image/png" href="{{ site.favicon_png }}" />
  <link rel="shortcut icon" type="image/x-icon" href="{{ site.favicon_ico }}">
  <link rel="canonical" href="{{ page.url | replace:'index.html','' | prepend: site.baseurl | prepend: site.url }}">
  <link rel='stylesheet' href='//fonts.googleapis.com/css?family=Open+Sans:400,600|Open+Sans+Condensed:300,700' type='text/css'>
  <link rel="stylesheet" href="{{ "/css/main.css" | prepend: site.baseurl }}">

  <title>{% if page.title %}{{ page.title | escape }}{% else %}{{ site.title | escape }}{% endif %}</title>
  <meta name="title" content="{% if page.title %}{{ page.title | escape }}{% else %}{{ site.title | escape }}{% endif %}">
  <meta name="google-site-verification" content="{{ site.webmaster.verification_code }}"/>
  <meta name="description" content="{{ site.description }}"/>
  <meta name="keywords" content="{{ site.keywords }}, {{ page.keywords }}"/>  
</head>
```

BTW, that's why I thought it's a generic problem... the header is parsed by your language link parser and with the I18 headers inside it makes sense that the links will be extended like all others...?

**Comment by untra** [comment]
AHA! `{% I18n_Headers %}` additionally needs the permanent address of your website. It needs to be used like this:
```html
{% I18n_Headers https://mywebsite.com %}
```
See my example [head.html](https://github.com/untra/polyglot/blob/site/_includes/head.html#L6)

**Comment by untra** [comment]
Okay, you have a two pages named _imprint_, an english one and a german one. Please make sure they both share the same permalink: `/imprint/`

If the german page has a frontmatter with
```yml
permalink: /de/imprint
lang: de
```
the permalink on the page will double up.

**Comment by cekageka** [comment]
You're right. Everything is fine with the url-parameter. The frontmatter is correct in my scenario, but if the url is not given the code seems to double the non-default alternate language tag. No idea why, but it's easy to reproduce. I will use an url, that's no problem.

But I have changed your code in i18n-headers.rb. I don't see a reason to double the url information form _config to any other layout-file, that's why I use this code to get the _config-url in the i18n-tag-code:

```
        def render(context)
          url = context.environments.first["site"]["url"]
```

Am I missing something or wouldn't that make sense anyway?

Thanks!

**Comment by untra** [comment]
That... does make more sense. Not every jekyll blog has a url parameter guarenteed, but it's a default field in the `_config.yml`, so why not?

I'll add that in as a default case when not specifying the url-parameter.

**Comment by cekageka** [comment]
Thanks for your support and thanks for the great plugin.

**Comment by atomobianco** [comment]
Hi there!
Having the same problem, if using only `{% I18n_Headers %} doubles the language url segment.
I also tried {% I18n_Headers site.url %} but it does not work.

**Comment by untra** [comment]
@atomobianco 
I think I know what is going on with that tag: The standard relativization is applying itself after the `I18n_Headers` runs. If it doesn't see a url, it just leaves relative urls (`<a href="/about">`) instead of site urls (`<a href=http://polyglot.untra.io/about/`)

try using `{% I18n_Headers http://yoursite.com %}` instead.

**Comment by atomobianco** [comment]
@untra 
That works, but it forces to have the site url declare in different places. Maybe leveraging the `{% I18n_Headers site.url %}` could help having one place for all. Is this a feature that would be difficult to integrate?

**Comment by untra** [comment]
yeah, I don't like it either :/
I am working on making a number of changes to the plugin, and I intend to update it with some new features I've been adding to support https://logrhythm.com . But more importantly, I'm waiting for [this PR to happen](https://github.com/jekyll/jekyll/pull/5344) within jekyll. This will give official parallel processing support to jekyll, which will define how polyglot should *really* do parallel site processing going forward. It will define the future of jekyll-polyglot, and probably jekyll too.

But yeah, I should fix that.
... or you can make a pr yourself? 😉

**Comment by atomobianco** [comment]
pardon me, you want to add cyber threats support – that's what I understand from your link ..
I will see if I am able to do a Pr, but no promises ;)

**Comment by niij** [comment]
I am having the same issue.  When using `{% I18n_Headers https://fullpathtomy.website/ %}` the headers work perfectly, which is my current fix.  When I just do `{% I18n_Headers %}` the language doubles up and it gives relative URL's, just like @cekageka is experiencing.  For reference, my url is set in my _config.yml

**Comment by niij** [comment]
@untra I just wanted to point out that this is happening on the demo site for this plugin, too.  [Here](http://polyglot.untra.io/fr/)

![image](https://user-images.githubusercontent.com/16074300/28263380-85bf31fa-6b21-11e7-8909-138dc04355e6.png)

**Comment by untra** [comment]
This looks to be fixed now (finally!) in the v1.3.0 release of polyglot. Check it out.
closing this issue now. Cheers!
